src/Microsoft.ML.Data/Transforms/ValueMappingTransform.cs

```diff
@@ -0,0 +1,515 @@
+            _columns = columns;
+        }
+
+        public override SchemaShape GetOutputSchema(SchemaShape inputSchema)
+        {
+            Host.CheckValue(inputSchema, nameof(inputSchema));
+
+            var resultDic = inputSchema.Columns.ToDictionary(x => x.Name);
+
+            var outputType = typeof(TValueType);
+            ColumnType outputColumnType = default;
+            if (outputType.IsGenericEx(typeof(VBuffer<>)))
+            {
+                Type vBufferType = outputType.GetGenericArguments()[0];
+                vBufferType.TryGetDataKind(out DataKind kind);
+                outputColumnType = new VectorType(PrimitiveType.FromKind(kind));
+            }
+            else
+            {
+                outputType.TryGetDataKind(out DataKind kind);
+                outputColumnType = PrimitiveType.FromKind(kind);
+            }
+
+            foreach (var (Input, Output) in _columns)
+            {
+                if (!inputSchema.TryFindColumn(Input, out var originalColumn))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", Input);
+
+                // Get the type from TOutputType
+                var col = new SchemaShape.Column(Output, originalColumn.Kind, outputColumnType, originalColumn.IsKey, originalColumn.Metadata);
+                resultDic[Output] = col;
+            }
+            return new SchemaShape(resultDic.Values);
+        }
+    }
+
+    public sealed class ValueMappingTransform<TKeyType, TValueType> : ValueMappingTransform
+    {
+        public ValueMappingTransform(IHostEnvironment env, IEnumerable<TKeyType> keys, IEnumerable<TValueType> values, (string Input, string Output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingTransform<TKeyType, TValueType>)),
+                  ConvertToDataView(env, keys, values), columns)
+        { }
+
+        public ValueMappingTransform(IHostEnvironment env, IEnumerable<TKeyType[]> keys, IEnumerable<TValueType> values, (string Input, string Output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingTransform<TKeyType, TValueType>)),
+                ConvertToDataView(env, keys, values), columns)
+        { }
+
+        public ValueMappingTransform(IHostEnvironment env, IEnumerable<TKeyType> keys, IEnumerable<TValueType[]> values, (string Input, string Output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingTransform<TKeyType, TValueType>)),
+                ConvertToDataView(env, keys, values), columns)
+        { }
+
+        public ValueMappingTransform(IHostEnvironment env, IEnumerable<TKeyType[]> keys, IEnumerable<TValueType[]> values, (string Input, string Output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingTransform<TKeyType, TValueType>)),
+                  ConvertToDataView(env, keys, values), columns)
+        { }
+
```

Comment:
Yes - since there are 4 different types of Estimator constructors, this trickled down to ValueMappingTransform and CreateToDataView.

---
In reply to: [236479820](https://github.com/dotnet/machinelearning/pull/1710#discussion_r236479820) [](ancestors = 236479820)